tst/EnergyPlus/unit/PVWatts.unit.cc

```diff
@@ -0,0 +1,283 @@
+		"   ** Severe  ** PVWatts: Invalid Geometry Type: ASDF",
+		"   **  Fatal  ** Errors found in getting PVWatts input",
+		"   ...Summary of Errors that led to program termination:",
+		"   ..... Reference severe error count=3",
+		"   ..... Last severe error=PVWatts: Invalid Geometry Type: ASDF"
+	});
+	EXPECT_TRUE( compare_err_stream( error_string, true ) );
+}
+
+TEST_F( EnergyPlusFixture, PVWattsGenerator_Calc )
+{
+	using namespace PVWatts;
+	// USA_AZ_Phoenix-Sky.Harbor.Intl.AP.722780_TMY3.epw
+	// 6/15 at 7am
+	DataGlobals::TimeStep = 1;
+	DataGlobals::TimeStepZone = 1.0;
+	DataHVACGlobals::TimeStepSys = 1.0;
+	DataGlobals::BeginTimeStepFlag = true;
+	DataGlobals::MinutesPerTimeStep = 60;
+	DataGlobals::NumOfTimeStepInHour = 1;
+	WeatherManager::AllocateWeatherData(); // gets us the albedo array initialized
+	DataEnvironment::Year = 1986;
+	DataEnvironment::Month = 6;
+	DataEnvironment::DayOfMonth = 15;
+	DataGlobals::HourOfDay = 8; // 8th hour of day, 7-8am
+	WeatherManager::WeatherFileLatitude = 33.45;
+	WeatherManager::WeatherFileLongitude = -111.98;
+	WeatherManager::WeatherFileTimeZone = -7;
+	DataEnvironment::BeamSolarRad = 728;
+	DataEnvironment::DifSolarRad = 70;
+	DataEnvironment::WindSpeed = 3.1;
+	DataEnvironment::OutDryBulbTemp = 31.7;
+
+	PVWattsGenerator pvw("PVWattsArrayA", 4000.0, ModuleType::STANDARD, ArrayType::FIXED_ROOF_MOUNTED);
+	pvw.setCellTemperature(30.345);
+	pvw.setPlaneOfArrayIrradiance(92.257);
+	pvw.calc();
+	Real64 generatorPower, generatorEnergy, thermalPower, thermalEnergy;
+	pvw.getResults(generatorPower, generatorEnergy, thermalPower, thermalEnergy);
+	EXPECT_DOUBLE_EQ(thermalPower, 0.0);
+	EXPECT_DOUBLE_EQ(thermalEnergy, 0.0);
+	EXPECT_NEAR(generatorPower, 884.137, 0.001);
+	EXPECT_NEAR(generatorEnergy, 884.137 * 60 * 60, 1);
+
+}
+
+TEST_F( EnergyPlusFixture, PVWattsInverter_Constructor )
+{
+	const std::string idfTxt = delimited_string({
+		"Version, 8.9;",
+		"ElectricLoadCenter:Distribution,",
+		"ELC1,",
+		"GeneratorList1,",
+		"Baseload,",
+		",",
+		",",
+		",",
+		"DirectCurrentWithInverter,",
+		"Inverter1;",
+		"ElectricLoadCenter:Inverter:PVWatts,",
+		"Inverter1,",
+		"1.10,",
+		"0.96;",
+		"ElectricLoadCenter:Generators,",
+		"GeneratorList1,",
+		"PVWattsArray1,",
+		"Generator:PVWatts,",
+		"1500,",
+		",",
+		",",
+		"PVWattsArray2,",
+		"Generator:PVWatts,",
+		"2500,",
+		",",
+		";",
+		"Generator:PVWatts,",
+		"PVWattsArray1,",
+		"5,",
+		"1500,",
+		"Standard,",
+		"OneAxis,",
+		",",
+		",",
+		",",
+		";",
+		"Generator:PVWatts,",
+		"PVWattsArray2,",
+		"5,",
+		"2500,",
+		"Standard,",
+		"OneAxis,",
+		",",
+		",",
+		",",
+		",",
+		",",
+		";"
+	});
+	ASSERT_FALSE(process_idf(idfTxt));
+	auto eplc(ElectPowerLoadCenter(1));
+	ASSERT_TRUE(eplc.inverterPresent);
+	EXPECT_DOUBLE_EQ(eplc.inverterObj->pvWattsDCCapacity(), 4000.0);
+	DataHVACGlobals::TimeStepSys = 1.0;
+	eplc.inverterObj->simulate(884.018);
+	EXPECT_NEAR(eplc.inverterObj->aCPowerOut(), 842.527, 0.001);
+
+}
```

Comment:
Overall the unit test approach looks great.  The calc test seems light, I'm curious how much of the code you are actually testing with it.  Looks like 47%.  You aren't covering the code for thin film, one axis backtracking, two axis, or surface geometry type.  Adding in some examples of those looks like it would push coverage of your new code up to 80%-ish.

Reply:
Fair enough.

Reply:
Frankly, if you can add a couple cases, to cover your variations, this thing will become a shining example of a unit tested new feature.  The bar will be set...